### DISCUSSION TITLE
Unlocking UE3 Framerate

### DISCUSSION THREAD

**Discussion by illusionyy** [xenia-canary/game-patches#16] [General]
Launching game with `--novsync` in later revisions of UE3 will allow for completely unlocked framerate.

Looking closer, there is one byte that catch interest.

![image](https://user-images.githubusercontent.com/37698908/117179835-850e1580-ae0e-11eb-95cc-182060f16c26.png)

```
82487d80 39 60 00 01     li         r11,0x1
```

I found that loadinig this with 0 at least on PS3, ignores:
VSync Flip Rate
MaxSmoothFramerate
(and possibly) bSmoothFramerate as well.

![image](https://user-images.githubusercontent.com/37698908/117180086-c3a3d000-ae0e-11eb-9afc-264b23b7bb9a.png)

This allows for completely unlocked fps without further settings change.

However, carrying this over to Xenia will only get rid of MaxSmoothFramerate. Vsync flip is still set to 2, manually setting it to 0 will only allow for 60FPS instead of 'Unlimited'. This can be resolved by setting vsync to false.

![image](https://user-images.githubusercontent.com/37698908/117180553-42007200-ae0f-11eb-8788-202fbec9fbb7.png)

**Comment by bomabomaboma** [comment]
I assume by flip rate you mean pre-rendered frames? If it's disabled, then I wonder what it's settling on....Are you experiencing input lag when it's disabled?

If so, that's very interesting. I had no idea that the --novsync command disabled the default amount of pre-rendered frames.

My only weariness about running a completely unlocked framerate is skipped frames, leading to an incongruence in delta time and inevitable scripting errors, where a game will break and possibly corrupt a save file. There can be a whole array of middleware in a title that can compound this problem, even above 30fps.

Smooth Frame Rate was a decent idea, but the practice of setting the max range at 62fps will promise framepacing issues when Xenia's vsync is disabled.

If we're looking to maximize performance, maintain delta timing, and get input lag as low as possible, there are two avenues:

- You can add a "--benchmark" cmd on top of --novsync, which will limit the game to one pre-rendered frame at a time.
- You can raise the min smoothed frame rate to ~45-50, adjust max smoothed frame rate to 59 (to prevent skipped frames on 60hz display) and set OneFrameThreadLag=False in Engine.ini [SystemSettings]....but that's a lot of memory patching.

The second option will certainly be more demanding, but it will deliver the best experience. (At least in the context of of guest rendering)
